beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,331 @@
+// a requirement outside of this package.
+func (dv *RODataColumnVerifier) SatisfyRequirement(req Requirement) {
+	dv.recordResult(req, nil)
+}
+
+func (dv *RODataColumnVerifier) recordResult(req Requirement, err *error) {
+	if err == nil || *err == nil {
+		dv.results.record(req, nil)
+		return
+	}
+	dv.results.record(req, *err)
+}
+
+// DataColumnIndexInBounds represents the follow spec verification:
+// [REJECT] The sidecar's index is consistent with NUMBER_OF_COLUMNS -- i.e. data_column_sidecar.index < NUMBER_OF_COLUMNS.
+func (dv *RODataColumnVerifier) DataColumnIndexInBounds() (err error) {
+	defer dv.recordResult(RequireBlobIndexInBounds, &err)
+	if dv.dataColumn.ColumnIndex >= fieldparams.NumberOfColumns {
+		log.WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("Sidecar index >= NUMBER_OF_COLUMNS")
+		return columnErrBuilder(ErrColumnIndexInvalid)
+	}
+	return nil
+}
+
+// NotFromFutureSlot represents the spec verification:
+// [IGNORE] The sidecar is not from a future slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+// -- i.e. validate that block_header.slot <= current_slot
+func (dv *RODataColumnVerifier) NotFromFutureSlot() (err error) {
+	defer dv.recordResult(RequireNotFromFutureSlot, &err)
+	if dv.clock.CurrentSlot() == dv.dataColumn.Slot() {
+		return nil
+	}
+	// earliestStart represents the time the slot starts, lowered by MAXIMUM_GOSSIP_CLOCK_DISPARITY.
+	// We lower the time by MAXIMUM_GOSSIP_CLOCK_DISPARITY in case system time is running slightly behind real time.
+	earliestStart := dv.clock.SlotStart(dv.dataColumn.Slot()).Add(-1 * params.BeaconConfig().MaximumGossipClockDisparityDuration())
+	// If the system time is still before earliestStart, we consider the column from a future slot and return an error.
+	if dv.clock.Now().Before(earliestStart) {
+		log.WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("sidecar slot is too far in the future")
+		return columnErrBuilder(ErrFromFutureSlot)
+	}
+	return nil
+}
+
+// SlotAboveFinalized represents the spec verification:
+// [IGNORE] The sidecar is from a slot greater than the latest finalized slot
+// -- i.e. validate that block_header.slot > compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)
+func (dv *RODataColumnVerifier) SlotAboveFinalized() (err error) {
+	defer dv.recordResult(RequireSlotAboveFinalized, &err)
+	fcp := dv.fc.FinalizedCheckpoint()
+	fSlot, err := slots.EpochStart(fcp.Epoch)
+	if err != nil {
+		return errors.Wrapf(columnErrBuilder(ErrSlotNotAfterFinalized), "error computing epoch start slot for finalized checkpoint (%d) %s", fcp.Epoch, err.Error())
+	}
+	if dv.dataColumn.Slot() <= fSlot {
+		log.WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("sidecar slot is not after finalized checkpoint")
+		return columnErrBuilder(ErrSlotNotAfterFinalized)
+	}
+	return nil
+}
+
+// ValidProposerSignature represents the spec verification:
+// [REJECT] The proposer signature of data_column_sidecar.signed_block_header,
```

Comment:
Should this rule be replaced by:

```
[REJECT] The proposer signature of sidecar.signed_block_header, is valid with respect to the block_header.proposer_index pubkey.
```
?

Reply:
Why this check is done so late in the pipeline, while in the spec it is done earlier?

Reply:
Signature checks are expensive which is why it has been deferred to the end, if an entitiy forges an invalid message we simply check for the other conditions first before verifying the proposer signature. Same way it was done for blobs